### ISSUE TITLE
Visual Studio runner not running all 1.9.2 tests

### ISSUE THREAD

**Issue by bradwilson** [xunit/visualstudio.xunit#71]
(Reported by @DanTup on Twitter)

Using the 1.9.2 source base and the 2.1.0-beta3 Visual Studio runner, this is the discovery list:

```
------ Discover test started ------
[xUnit.net 00:00:00.0426065] Discovery starting: test.xunit.console (name display = ClassAndMethod)
[xUnit.net 00:00:00.2630928] Discovery finished: test.xunit.console (52 tests)
[xUnit.net 00:00:00.0377609] Discovery starting: test.xunit.extensions (name display = ClassAndMethod)
[xUnit.net 00:00:00.2666569] Discovery finished: test.xunit.extensions (66 tests)
[xUnit.net 00:00:00.0355687] Discovery starting: test.xunit.gui (name display = ClassAndMethod)
[xUnit.net 00:00:00.2424575] Discovery finished: test.xunit.gui (13 tests)
[xUnit.net 00:00:00.0364888] Discovery starting: test.xunit.runner.msbuild (name display = ClassAndMethod)
[xUnit.net 00:00:00.2314477] Discovery finished: test.xunit.runner.msbuild (3 tests)
[xUnit.net 00:00:00.0375649] Discovery starting: test.xunit.runner.tdnet (name display = ClassAndMethod)
[xUnit.net 00:00:00.2360481] Discovery finished: test.xunit.runner.tdnet (10 tests)
[xUnit.net 00:00:00.0390550] Discovery starting: test.xunit.runner.utility (name display = ClassAndMethod)
[xUnit.net 00:00:00.3297654] Discovery finished: test.xunit.runner.utility (178 tests)
[xUnit.net 00:00:00.0507593] Discovery starting: test.xunit (name display = ClassAndMethod)
[xUnit.net 00:00:00.4388930] Discovery finished: test.xunit (379 tests)
========== Discover test finished: 701 found (0:00:02.7418674) ==========
```

And this is the run:

```
------ Run test started ------
[xUnit.net 00:00:00.0152959] Execution starting: test.xunit.console (method display = ClassAndMethod, parallel test collections = True, max threads = 8)
[xUnit.net 00:00:00.5246242] Execution finished: test.xunit.console
[xUnit.net 00:00:00.0153263] Execution starting: test.xunit.extensions (method display = ClassAndMethod, parallel test collections = True, max threads = 8)
[xUnit.net 00:00:00.0095983] Execution starting: test.xunit.gui (method display = ClassAndMethod, parallel test collections = True, max threads = 8)
[xUnit.net 00:00:00.3927170] Execution finished: test.xunit.gui
[xUnit.net 00:00:00.0106905] Execution starting: test.xunit.runner.msbuild (method display = ClassAndMethod, parallel test collections = True, max threads = 8)
[xUnit.net 00:00:01.0433210] Execution finished: test.xunit.runner.msbuild
[xUnit.net 00:00:00.0101530] Execution starting: test.xunit.runner.tdnet (method display = ClassAndMethod, parallel test collections = True, max threads = 8)
[xUnit.net 00:00:00.4181491] Execution finished: test.xunit.runner.tdnet
[xUnit.net 00:00:00.0242951] Execution starting: test.xunit.runner.utility (method display = ClassAndMethod, parallel test collections = True, max threads = 8)
[xUnit.net 00:00:09.4441385] Execution finished: test.xunit.runner.utility
[xUnit.net 00:00:00.0371942] Execution starting: test.xunit (method display = ClassAndMethod, parallel test collections = True, max threads = 8)
========== Run test finished: 291 run (0:00:23.4857307) ==========
```

**Comment by bradwilson** [comment]
Fixed in b2d753d5a29c17096289ca1da57e27347d41935a

**Comment by DanTup** [comment]
Great; thanks! Is it likely there will be a new NuGet with this fix prior to VS2015 RTM?

I'm also curious, what was the issue? I scanned the change, but couldn't really figure out which bit was the fix.

**Comment by bradwilson** [comment]
It will be in 2.1 Beta 4 for sure, which I'd guess comes out around the same time as 2015 RTM.

I'm actually considering cherry-picking this change and putting it into a 2.0.1 release, since it fixes a pretty significant issue with people running 1.x tests.

**Comment by bradwilson** [comment]
The bug on our side was that we should've been using `ITestCaseStarting` and `ITestCaseFinished` when calling `RecordStart` and `RecordEnd` with Visual Studio. This appeared to have been a benign mistake on our part. You can see the changes here:

https://github.com/xunit/visualstudio.xunit/commit/b2d753d5a29c17096289ca1da57e27347d41935a#diff-d48eb63ec121f45ca6ec37f588861195R107
https://github.com/xunit/visualstudio.xunit/commit/b2d753d5a29c17096289ca1da57e27347d41935a#diff-d48eb63ec121f45ca6ec37f588861195R118

The bug on the Visual Studio side (in 2015 only) which actually caused the issue is what's referenced in #72. Our code was not very defensive, and these unexpected exceptions were causing reporting issues to the end user. We fixed that by wrapping all calls to VS with try/catch that will report the issue:

https://github.com/xunit/visualstudio.xunit/commit/b2d753d5a29c17096289ca1da57e27347d41935a#diff-d48eb63ec121f45ca6ec37f588861195R47

So, on VS 2015, all theories will more than 1 data row will cause these exceptions to be surfaced to the end user in the Tests output pane. I've e-mailed someone on the VS Test team to see if this can get addressed before VS 2015 RTM, but with the date just a few weeks away, I'm guessing we will have missed the window.

I don't know whether the issue only affects Visual Studio, or whether it also affects Team Build, so we may need to suppress these messages (or ratchet them down to warnings) if the issue cannot be resolved by 2015 RTM.

**Comment by DanTup** [comment]
Thanks for the info. I'll keep my eyes out for a 2.0.1, would prefer to stay on stable versions if possible.

**Comment by bradwilson** [comment]
I just pushed out a 2.0.1 RC. If you could give it a try and make sure your issues are resolved, that'd be great. :)

http://xunit.github.io/release-notes/2015-07-04.html

**Comment by DanTup** [comment]
Will test tomorrow morning when back at the PC. Thanks!

**Comment by DanTup** [comment]
Problem appears to be resolved in 2.0.1 RC; all tests running as before. Thanks!

Btw, would be handy to have references back to the issues in the [release notes](http://xunit.github.io/release-notes/2015-07-04.html).
